### PR TITLE
[KT-15695] String.nullIfEmpty

### PR DESCRIPTION
**Pull request by MGaetan89**
This PR adds the following extensions functions:
- `Array.nullIfEmpty()`
- `CharSequence.nullIfEmpty()`
- `Collection.nullIfEmpty()`
- `String.nullIfEmpty()`

It also fixes a couple of warning when running `./gradlew :tools:kotlin-stdlib-gen:run`.

Related issue: https://youtrack.jetbrains.com/issue/KT-15695

### CODE REVIEW

**Comment by MGaetan89** [comment]
I'm using the same sample for every kind of `Array`. Do you prefer a dedicated sample for each kind?

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt [hunk 1]
```diff
@@ -187,5 +185,26 @@ fun arrays(): List<GenericFunction> {
...
+            else if (f == Collections)
+                "@sample samples.collections.Collections.nullIfEmpty"
+            else
+                ""
+            }
+            """
+        }
+        returns("SELF?")
+        body {
+            "return if (this == null || this.isEmpty()) null else this"
```
Comment (ilya-g) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: `this == null` 
^ expression is always false

File: libraries/tools/kotlin-stdlib-gen/src/templates/Elements.kt [hunk 2]
```diff
@@ -14,7 +14,7 @@ fun elements(): List<GenericFunction> {
         doc { f -> "Returns `true` if [element] is found in the ${f.collection}." }
         typeParam("@kotlin.internal.OnlyInputTypes T")
         returns("Boolean")
-        body(Iterables) { f ->
+        body(Iterables) {
```
Comment (ilya-g) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Elements.kt]: Please extract unrelated changes, such as formatting improvement in a separate PR.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 1 above)
Comment (MGaetan89) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: You're right. I mixed it with the `String`/`CharSequence` implementation.
Should I unify both to apply on non-nullable type?
I will also revert the formatting changes.

**Comment by MGaetan89** [comment]
@ilya-g I've remove all `this == null` checks, and made `nullIfEmpty()` extension available only on non-nullable `CharSequence` and `String`.

**Comment by ilya-g** [comment]
> I'm using the same sample for every kind of Array. 

That's enough to give the usage example.

**Comment by MGaetan89** [comment]
Is there something else to do for this to get merged?

**Comment by ilya-g** [comment]
It would be nice to cover this with some tests, despite the fact they might sound trivial.
To get this merged, we should first discuss this function on a design meeting, and it could take some time to complete, so please be patient.

**Comment by pun-ky** [comment]
Adding these methods is surprising for me because Kotlin rather encourages to use non-null types so that having that methods seems for me a little bit not a way to go 😉

**Comment by ilya-g** [comment]
@pun-ky While Kotlin encourages working with non-nullable values it also provides expressive ways of dealing with nullable ones, e.g. safe call `?.` and elvis `?:` operators. It might be advantageous to turn an empty collection into `null` and then use these operators:

```
val keywords = getFilterKeywords().nullIfEmpty() ?: return
...
```

So there's nothing wrong with this feature request. 
Though in Kotlin 1.3 the issue [KT-15695](https://youtrack.jetbrains.com/issue/KT-15695) was implemented in a different way: as a more general purpose function `ifEmpty`, so that `ifEmpty { null }` could be used instead of `nullOfEmpty()`.

**Comment by pun-ky** [comment]
Seems legit 👍 especially when ifEmpty is accepting value provider / lambda. Cool!
